### PR TITLE
Allow creating Clang compilation database for third-party dependencies

### PR DESCRIPTION
**Pull request by mbautin**

### CODE REVIEW

File: python/yugabyte_db_thirdparty/compile_commands.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+from yugabyte_db_thirdparty import util
+from yugabyte_db_thirdparty.custom_logging import log
+
+
+COMPILE_COMMAND_FILE_SUFFIX = '.compile_command.json'
+
+TMP_DIR_ENV_VAR_NAME = 'YB_THIRDPARTY_COMPILE_COMMANDS_TMP_DIR'
+
+SOURCE_FILE_SUFFIXES = ('.h', '.hpp', '.cc', '.c', '.cxx')
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compile_commands.py:36]: Unused?

File: python/yugabyte_db_thirdparty/compile_commands.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    return os.path.join(
+        get_compile_commands_dir(build_dir), 'compile_commands%s.json' % ('_raw' if raw else ''))
+
+
+def should_include_compile_command(compile_command: Dict[str, Union[str, List[str]]]) -> bool:
+    file_path = cast(str, compile_command['file'])
+    file_basename = os.path.basename(file_path)
+    dir_path = cast(str, compile_command['directory'])
+    dir_basename = os.path.basename(dir_path)
+    return not file_basename == 'conftest.c' and dir_basename not in ['conftest', 'conftest.dir']
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compile_commands.py:90]: What is conftest and why do we have to handle it separately? (Add a comment?)

File: python/yugabyte_db_thirdparty/compiler_wrapper.py [hunk 3]
```diff
@@ -16,14 +16,22 @@
...
 from yugabyte_db_thirdparty.util import shlex_join, is_shared_library_name
 from yugabyte_db_thirdparty.constants import (
     COMPILER_WRAPPER_ENV_VAR_NAME_LD_FLAGS_TO_APPEND,
     COMPILER_WRAPPER_ENV_VAR_NAME_LD_FLAGS_TO_REMOVE,
 )
+from yugabyte_db_thirdparty.util import mkdir_p
+from yugabyte_db_thirdparty import compile_commands
+
+
+C_CXX_SUFFIXES = ('.c', '.cc', '.cxx')
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compiler_wrapper.py:34]: .cpp too?

File: python/yugabyte_db_thirdparty/compiler_wrapper.py [hunk 4]
```diff
@@ -108,22 +134,35 @@ def run(self) -> None:
...
+                ]
+                if len(input_file_candidates) != 1:
+                    sys.stderr.write(
+                        f"Could not determine input file name for compiler invocation, will omit "
+                        f"from compile commands. Input file candidates: {input_file_candidates}, "
+                        f"command line: {self._get_compiler_command_str()}"
+                    )
+                    generate_compile_command_file = False
+
+            # Perform preprocessing only to ensure we are only using include files from allowed
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compiler_wrapper.py:160]: Nit: only to ensure -> to ensure

File: python/yugabyte_db_thirdparty/compile_commands.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    compile_commands = util.read_json_file(compile_commands_path_raw)
+    new_compile_commands = []
+    for compile_command in compile_commands:
+        new_compile_commands.append(
+            rewrite_compile_command(
+                cast(Dict[str, Union[str, List[str]]], compile_command),
+                bazel_path_mapping))
+
+    compile_commands_path = get_final_compile_commands_path(build_dir, raw=False)
+    with open(compile_commands_path, 'w') as compile_commands_file:
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compile_commands.py:213]: Use `write_json_file`?

File: python/yugabyte_db_thirdparty/compile_commands.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+    # Merge the existing compile commands with the new ones.
+    for existing_compile_command in existing_compile_commands:
+        if existing_compile_command['file'] not in new_files:
+            compile_commands.append(existing_compile_command)
+
+    compile_commands = filter_compile_commands(compile_commands)
+
+    compile_commands = sorted(compile_commands, key=lambda c: c['file'])
+    with open(aggregated_path_raw, 'w') as compile_commands_file:
```
Comment (SrivastavaAnubhav) [inline python/yugabyte_db_thirdparty/compile_commands.py:145]: Use `write_json_file`?

**Review by SrivastavaAnubhav — APPROVED** [review]
Minor comments

File: python/yugabyte_db_thirdparty/builder.py [hunk 7]
```diff
@@ -260,12 +266,14 @@ def parse_args(self) -> None:
...
             activate_devtoolset(self.args.devtoolset)
+
+        use_compiler_wrapper = self.args.use_compiler_wrapper
         self.compiler_choice = CompilerChoice(
             compiler_family=compiler_family,
             compiler_prefix=compiler_prefix,
             compiler_suffix=self.args.compiler_suffix,
             devtoolset=self.args.devtoolset,
-            use_compiler_wrapper=False,
+            use_compiler_wrapper=use_compiler_wrapper,  # We will change it later.
```
[the remark this answers is not in the document]
Comment (mbautin) [reply python/yugabyte_db_thirdparty/builder.py]: Removed.

File: python/yugabyte_db_thirdparty/builder.py [hunk 8]
```diff
@@ -858,19 +892,29 @@ def build_one_build_type(self, build_type: str) -> None:
...
-                        f"should_build={should_build}, "
-                        f"should_rebuild={should_rebuild}.")
+        dependencies_matching_group = [
+            dep for dep in self.selected_dependencies if dep.build_group == build_group
+        ]
+        for dep in dependencies_matching_group:
+            self.perform_pre_build_steps(dep)
+
+        self.bazel_path_mapping = {}
+        for dep in self.dependencies:
```
[the remark this answers is not in the document]
Comment (mbautin) [reply python/yugabyte_db_thirdparty/builder.py]: Moved this to the top-level `run` function.

File: python/yugabyte_db_thirdparty/builder.py [hunk 9]
```diff
@@ -1224,11 +1274,38 @@ def build_dependency(self, dep: Dependency, only_process_flags: bool = False) ->
...
+                compile_commands.postprocess_compile_commands(
+                    build_dir, self.bazel_path_mapping, clang_toolchain_dir)
+                return
+
+            with PushDir(build_dir):
+                with EnvVarContext(**env_vars):
+                    write_env_vars(DEPENDENCY_ENV_FILE_NAME)
+                    log("PATH=%s" % os.getenv('PATH'))
+                    dep.build(self)
+            if compile_commands_tmp_dir is not None:
```
[the remark this answers is not in the document]
Comment (mbautin) [reply python/yugabyte_db_thirdparty/builder.py:1342]: Yes, but we set the `compile_commands_tmp_dir` variable to the return value of that function only under the following condition:`self.args.compile_commands and self.build_type.is_sanitizer()`.

File: python/yugabyte_db_thirdparty/builder.py [hunk 10]
```diff
@@ -1224,11 +1274,38 @@ def build_dependency(self, dep: Dependency, only_process_flags: bool = False) ->
...
+                compile_commands.aggregate_compile_commands(
+                    compile_commands_tmp_dir, build_dir, self.bazel_path_mapping,
+                    clang_toolchain_dir)
+        finally:
+            if compile_commands_tmp_dir is not None:
```
[the remark this answers is not in the document]
Comment (mbautin) [reply python/yugabyte_db_thirdparty/builder.py:1347]: Ditto

File: python/yugabyte_db_thirdparty/compile_commands.py (hunk 1 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compile_commands.py:36]: Removed.

File: python/yugabyte_db_thirdparty/compiler_wrapper.py (hunk 3 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compiler_wrapper.py:34]: Done.

File: python/yugabyte_db_thirdparty/compile_commands.py (hunk 2 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compile_commands.py:92]: Added a comment.

File: python/yugabyte_db_thirdparty/compile_commands.py (hunk 5 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compile_commands.py:327]: Done.

File: python/yugabyte_db_thirdparty/compile_commands.py (hunk 6 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compile_commands.py:147]: Done.

File: python/yugabyte_db_thirdparty/builder.py [hunk 11]
```diff
@@ -170,6 +172,10 @@ class Builder(BuilderInterface):
     lto_type: Optional[str]
     selected_dependencies: List[Dependency]
 
+    # Mapping from e.g. com_google_tcmalloc or com_google_absl to the corresponiding build
```
[the remark this answers is not in the document]
Comment (mbautin) [reply python/yugabyte_db_thirdparty/builder.py:175]: Fixed.

File: python/yugabyte_db_thirdparty/compiler_wrapper.py (hunk 4 above)
Comment (mbautin) [reply python/yugabyte_db_thirdparty/compiler_wrapper.py:160]: Fixed.

**Pull request merged**